src/redirects.js

```diff
@@ -0,0 +1,79 @@
+import { getConfig } from 'src/config'
+
+const REDIRECT_DEFAULTS = {
+  basePath: true,
+  permanent: false,
+}
+
+const getDestination = ({ ctx, AuthUser, redirectDestination }) => {
+  if (typeof redirectDestination === 'function') {
+    const destination = redirectDestination({ ctx, AuthUser })
+    return getDestination({ ctx, AuthUser, redirectDestination: destination })
+  }
+
+  if (typeof redirectDestination === 'string') {
+    return {
+      ...REDIRECT_DEFAULTS,
+      destination: redirectDestination,
+    }
+  }
+
+  if (typeof redirectDestination === 'object') {
+    return {
+      ...REDIRECT_DEFAULTS,
+      ...redirectDestination,
+    }
+  }
+
+  return null
+}
+
+const throwWhenInvalid = (prop, redirectURL) => {
+  const isValid =
+    redirectURL &&
+    (typeof redirectURL === 'string' ||
+      (typeof redirectURL && 'destination' in redirectURL))
+
+  if (!isValid) {
+    throw new Error(
+      `The "${prop}" must be set to a non-empty string, an object literal containing "destination", or a function that returns either.`
+    )
+  }
+
+  return redirectURL
+}
+
+const getRedirectByUrlConfigName = (redirectConfig) => {
+  const { redirectConfigName, options, ctx, AuthUser } = redirectConfig
+  return throwWhenInvalid(
+    redirectConfigName,
+    getDestination({
+      ctx,
+      AuthUser,
+      redirectDestination:
+        options && options[redirectConfigName]
+          ? options[redirectConfigName]
+          : getConfig()[redirectConfigName],
+    })
+  )
+}
+
+export const getRedirectToLoginDestination = ({
+  options,
+  AuthUser,
+  ctx,
+} = {}) =>
+  getRedirectByUrlConfigName({
```

Comment:
Suggestion for clarity: get the default URL from the config here, along with the urlSetting, and pass those—or even just pass the relevant one (config value if urlSetting is not defined). That would be easier to reason about rather than relying on config key names.

Reply:
The only side effect here is that the error message relies knowing the prop name, but maybe we can move this validation up so where we don't need to reason about a config setting name.

Reply:
But, I mean we could rethink the error message and remove the complexity.

Reply:
@kmjennison What are your thoughts on updating this error message for the new context of being within the redirect module? Something like `The "redirectDestination" must be set to a non-empty string ...`? The caller will know what it means because it needs to be explicitly set in the props for the function.

Reply:
@offgriddev I think it's still fine to pass the "redirectConfigName" value for use only in error messages.

Reply:
Sounds good. I'll work this in today.